File: python/leetcode/314.py
class TreeNode():
    def __init__(self, val):
        self.val = val
        self.right = None
        self.left = None

class Solution:
    def verticalOrder(self, root):
        if not root: return []
        nodes = collections.defaultdict(list)
        queues = [(root, 0)]
        for node, col in queues:
            if node:
                nodes[col].append(node.val)
                queues.append((node.left, col - 1))
                queues.append((node.right, col + 1))
        return [nodes[col] for col in sorted(nodes)]

File: cpp/leetcode/314.cpp
class Solution {
public:
    vector<vector<int>> verticalOrder(TreeNode* root) {
        if(!root) return {};
        vector<vector<int>> res;
        map<int, vector<int>> hmap;
        queue<pair<int, TreeNode*>> q;
        q.push(make_pair(0, root));
        while(!q.empty()){
            int n = q.size();
            for(int i = 0; i < n; i++){
                int level = q.front().first;
                TreeNode * node = q.front().second;
                q.pop();
                hmap[level].push_back(node->val);
                if(node->left) q.push(make_pair(level - 1, node->left));
                if(node->right) q.push(make_pair(level + 1, node->right));
            }
        }
        for(auto it : hmap) res.push_back(it.second);
        return res;
    }
};